### PR TITLE
change osr-def syntax

### PR DESCRIPTION
**Pull request by o-**
use the same syntax for references as in the call case

Also adds some tests

### CODE REVIEW

**Comment by mhyee** [comment]
I honestly have no intuition/preference between `mut x <- y`/`mut x = e` or `mut x <- e`/`mut x = y` (which is probably a reason to change it). But I agree consistency with the call syntax is probably better.

For the tests, could you also add a case where we have a non-variable expression?

Also, don't forget to update the semantics doc.

**Comment by o-** [comment]
> Also, don't forget to update the semantics doc.

riight. done

**Comment by o-** [comment]
> For the tests, could you also add a case where we have a non-variable expression?

Don't understand. There is a test with const. I guess you meant something diffrent.

Update: sorry, get it

File: parser.mly [hunk 1]
```diff
@@ -110,9 +110,9 @@ osr_def:
     { Osr_const (x, e) }
 | MUT x=variable
     { Osr_mut_undef x }
-| MUT x=variable EQUAL y=variable
+| MUT x=variable EQUAL AMPERSAND y=variable
     { Osr_mut_ref (x, y) }
-| MUT x=variable LEFTARROW e=expression
+| MUT x=variable EQUAL e=expression
```
Comment (gasche) [inline parser.mly:115]: Do we need to update the parser? I wonder if some error messages mention this syntax.

**Review by gasche — APPROVED** [review]
Ready for merge -- maybe see the parser.messages question.

File: parser.mly (hunk 1 above)
Comment (o-) [reply parser.mly:115]: yeah, I think they are out of sync anyways :(

File: doc/ir-semantics.text [hunk 2]
```diff
@@ -166,17 +175,20 @@ Update (partial) function, returns an H:
...
+         H' := H
          E' := E
      and if v = true then
          Df ∈ dom P
          Df  = Function f^ (formal*)
          I' := P(Df, lᵥ^)
          l' := l^
+         M  := (materialize P H E osr-map)*
+         H' := H ∪ rng(M)
          E' := (osr-env P H E osr-map)*
```
Comment (mhyee) [inline doc/ir-semantics.text:192]: `M` was added to `osr-env`, so this should be: `E' := (osr-env P H E M osr-map)*`

File: doc/ir-semantics.text [hunk 3]
```diff
@@ -110,10 +111,18 @@ Update (partial) function, returns an H:
...
+  materialize P H E (mut x = e) = (x ↦ (a ↦ v))
+    where a - fresh
+      and v := eval P H E e
+  materialize P H E _           = ()
+
+  osr-env P H E M (const x = e) = (x ↦ val v)
+    where v := eval P H E e
+  osr-env P H E M (mut x)       = (x ↦ ⊥)
+  osr-env P H E M (mut x = &y)  = (x ↦ ref a)
     where a = E(y)
```
Comment (mhyee) [inline doc/ir-semantics.text:123]: Looks like the notation `E ∋ (y ↦ ref a)` is used more often.

File: doc/ir-semantics.text [hunk 4]
```diff
@@ -110,10 +111,18 @@ Update (partial) function, returns an H:
...
+  osr-env P H E M (mut x = e)   = (x ↦ ref e)
```
Comment (mhyee) [inline doc/ir-semantics.text:124]: Does `e` need to be evaluated?

Comment (mhyee) [reply doc/ir-semantics.text:124]: Oh, I think I get it. `e` is evaluated by `materialize`.

But then it should be `(x ↦ ref a)`, right?

File: doc/ir-semantics.text [hunk 5]
```diff
@@ -110,10 +111,18 @@ Update (partial) function, returns an H:
...
+    where (a ↦ _) = M(x)
```
Comment (mhyee) [inline doc/ir-semantics.text:125]: Likewise, I think it would be more consistent to write `M ∋ (x ↦ (a ↦ _))`

File: doc/ir-semantics.text [hunk 6]
```diff
@@ -61,12 +61,13 @@ v :=     values
...
 | mut x
-| mut x = x
+| mut x = &x
+| mut x = e
 
 # Natural operational semantics
 
 E ::= (x ↦ val v | x ↦ ref a | x ↦ ⊥)*  lexical environment
-H ::= (l ↦ v)                           mutable heap
+H ::= (a ↦ v)                           mutable heap
```
Comment (mhyee) [inline doc/ir-semantics.text:70]: While I'm at it, nitpick at how "mutable heap" doesn't line up with "lexical environment."

File: doc/ir-semantics.text [hunk 7]
```diff
@@ -166,17 +175,20 @@ Update (partial) function, returns an H:
...
          l' := l
+         H' := H
          E' := E
      and if v = true then
          Df ∈ dom P
          Df  = Function f^ (formal*)
          I' := P(Df, lᵥ^)
          l' := l^
+         M  := (materialize P H E osr-map)*
+         H' := H ∪ rng(M)
```
Comment (mhyee) [inline doc/ir-semantics.text:191]: It took me a while to realize that `rng` was range. I'd prefer it to be explicitly written out.

**Comment by mhyee** [comment]
I had some more nitpicking/stupid questions. Sorry I couldn't get them in before the merge, I was in class.

**Comment by gasche** [comment]
Sorry for merging too fast! If you know what you would prefer to see instead for each of these changes, feel free to do them yourself and send a followup/cleanup PR.

**Comment by mhyee** [comment]
Sure, will do so.

**Comment by o-** [comment]
> Sure, will do so.

I can fix it too, but if you want, go ahead

File: doc/ir-semantics.text (hunk 6 above)
Comment (mhyee) [reply doc/ir-semantics.text:70]: Turns out the columns line up, but some symbols aren't the same width? Oh well.

**Pull request merged**